b2sdk/bucket.py

```diff
@@ -581,6 +586,111 @@ def upload_local_file(
+        legal_hold: Optional[LegalHold] = None,
+        min_part_size: Optional[int] = None,
+        max_part_size: Optional[int] = None,
+        large_file_sha1: Optional[Sha1HexDigest] = None,
+        buffers_count: int = 2,
+        read_size: int = 8192,
+        unused_buffer_timeout_seconds: float = 3600.0,
+    ):
+        """
+        Upload an unbound file-like read-only object to a B2 file.
+
+        It is assumed that this object is streamed like stdin or socket, and the size is not known apriori.
+        It is up to caller to ensure that this object is open and available through the whole streaming process.
+
+        If stdin is to be passed, consider opening it in binary mode, if possible on the platform:
+
+        .. code-block:: python
+
+            with open(sys.stdin.fileno(), mode='rb', buffering=min_part_size, closefd=False) as source:
+                bucket.upload_unbound_stream(source, 'target-file')
+
+        For platforms without file descriptors, one can use the following:
+
+        .. code-block:: python
+
+            bucket.upload_unbound_stream(sys.stdin.buffer, 'target-file')
+
+        but note that buffering in this case is depending on interpreter mode.
+
+        ``min_part_size``, ``recommended_upload_part_size`` and ``max_part_size`` should all be greater than 5MB.
+        This is the minimal size of a chunk that is to be uploaded.
+
+        ``buffers_count`` describes a desired number of buffers that are to be used. Minimal amount is two, as we need
+        to determine the method of uploading this stream (if there's only a single buffer we send it as a normal file,
+        if there are at least two – as a large file).
+        In rare cases, namely when the whole buffer was sent, but there was an error during sending of last bytes
+        and a retry was issued, another buffer (above the aforementioned limit) will be allocated.
+
+        :param file-like-object read_only_object: any object containing read method accepting size of the read
+        :param str file_name: a file name of the new B2 file
+        :param str,None content_type: the MIME type, or ``None`` to accept the default based on file extension of the B2 file name
+        :param dict,None file_info: a file info to store with the file or ``None`` to not store anything
+        :param b2sdk.v2.AbstractProgressListener,None progress_listener: a progress listener object to use, or ``None`` to not report progress
+        :param b2sdk.v2.EncryptionSetting encryption: encryption settings (``None`` if unknown)
+        :param b2sdk.v2.FileRetentionSetting file_retention: file retention setting
+        :param bool legal_hold: legal hold setting
+        :param int min_part_size: a minimum size of a part
+        :param int,None recommended_upload_part_size: the recommended part size to use for uploading local sources
+                        or ``None`` to determine automatically; also size of the read buffer and size of the chunks
```

Comment:
I've added `buffer_size` as an explicit parameter and added a information that, if left unset, it'll default to a value of `recommended_upload_part_size`.

`buffer_size` is (currently) equal to the size of the data being sent to B2. This is an implementation detail connected to way the `WriteIntent` is passed further (that is – lack of any preprocessing on that regard). Providing another value for the chunks being sent requires additional set of checks between `min` and `max` upload sizes, `buffer_size` and `chunk_size` to ensure that it actually "can" be uploaded (how many buffers fit in a chunk, do we have enough buffers to fill in at least one chunk, is chunk small/large enough to be uploaded to B2; would be nice if, in case of error, a "similar but working" configuration was returned, explanations of all the logic for the user) – and sadly I didn't find enough justification for all that additional work. In current model we have a rough estimate of the memory used by the process and that calculation is as simple as `buffer_count * buffer_size`.

This is why I didn't want to provide another value for `buffer_size` nor `chunk_size` (the latter is not even configurable right now).